model/atmosphere/subgrid_scale_physics/muphys/tests/graupel/test_graupel.py

```diff
@@ -0,0 +1,408 @@
+
+
+grid_geometries = {}
+
+
+def get_grid_geometry(backend, grid_file) -> geometry.GridGeometry:
+    def construct_decomposition_info(grid: icon.IconGrid) -> definitions.DecompositionInfo:
+        edge_indices = alloc.allocate_indices(dims.EdgeDim, grid)
+        owner_mask = np.ones((grid.num_edges,), dtype=bool)
+        decomposition_info = definitions.DecompositionInfo(klevels=grid.num_levels)
+        decomposition_info.with_dimension(dims.EdgeDim, edge_indices.ndarray, owner_mask)
+        return decomposition_info
+
+    def construct_grid_geometry(grid_file: str):
+        gm = utils.run_grid_manager(grid_file)
+        grid = gm.grid
+        decomposition_info = construct_decomposition_info(grid)
+        geometry_source = geometry.GridGeometry(
+            grid, decomposition_info, backend, gm.coordinates, gm.geometry, attrs.attrs
+        )
+        return geometry_source
+
+    if not grid_geometries.get(grid_file):
+        grid_geometries[grid_file] = construct_grid_geometry(grid_file)
+    return grid_geometries[grid_file]
+
+
+def test_geometry_raises_for_unknown_field(backend):
+    geometry = get_grid_geometry(backend, dt_utils.R02B04_GLOBAL)
+    with pytest.raises(ValueError) as e:
+        geometry.get("foo")
+        assert "'foo'" in e.value
+        assert "'GridGeometry'" in e.value
+
+
+@pytest.mark.parametrize(
+    "grid_file, experiment, rtol",
+    [
+        (dt_utils.REGIONAL_EXPERIMENT, dt_utils.REGIONAL_EXPERIMENT, 5e-9),
+        (dt_utils.R02B04_GLOBAL, dt_utils.GLOBAL_EXPERIMENT, 3e-12),
+    ],
+)
+@pytest.mark.datatest
+def test_edge_control_area(backend, grid_savepoint, grid_file, rtol):
+    expected = grid_savepoint.edge_areas()
+    geometry_source = get_grid_geometry(backend, grid_file)
+    result = geometry_source.get(attrs.EDGE_AREA)
+    assert helpers.dallclose(expected.ndarray, result.ndarray, rtol)
+
+
+@pytest.mark.parametrize(
+    "grid_file, experiment",
+    [
+        (dt_utils.REGIONAL_EXPERIMENT, dt_utils.REGIONAL_EXPERIMENT),
+        (dt_utils.R02B04_GLOBAL, dt_utils.GLOBAL_EXPERIMENT),
+    ],
+)
+@pytest.mark.datatest
+def test_coriolis_parameter(backend, grid_savepoint, grid_file):
+    geometry_source = get_grid_geometry(backend, grid_file)
+    expected = grid_savepoint.f_e()
+
+    result = geometry_source.get(attrs.CORIOLIS_PARAMETER)
+    assert helpers.dallclose(expected.ndarray, result.ndarray)
+
+
+@pytest.mark.parametrize(
+    "grid_file, experiment, rtol",
+    [
+        (dt_utils.REGIONAL_EXPERIMENT, dt_utils.REGIONAL_EXPERIMENT, 1e-9),
+        (dt_utils.R02B04_GLOBAL, dt_utils.GLOBAL_EXPERIMENT, 1e-12),
+    ],
+)
+@pytest.mark.datatest
+def test_compute_edge_length(backend, grid_savepoint, grid_file, rtol):
+    geometry_source = get_grid_geometry(backend, grid_file)
+    expected = grid_savepoint.primal_edge_length()
+    result = geometry_source.get(attrs.EDGE_LENGTH)
+    assert helpers.dallclose(result.ndarray, expected.ndarray, rtol=rtol)
```

Comment:
```suggestion
    assert helpers.dallclose(result.asnumpy(), expected.asnumpy(), rtol=rtol)
```

Reply:
same for the others

Reply:
This is the test program for muphys-cpp with Python bindings.  I will delete this file.